### DISCUSSION TITLE
Enhance FastAPI's Route Ordering to Prevent Unintended Matching

### DISCUSSION THREAD

**Discussion by linkedlist771** [fastapi/fastapi#11578] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI() 

@app.get("/users/{user_id}")
async def read_user(user_id: str):
    return {"user_id": user_id}

@app.get("/users/me")
async def read_current_user():
    return {"user_id": "the current user"}
```

### Description

When defining routes in FastAPI, the order matters. In the code above, a request to `/users/me` will actually be handled by the `/users/{user_id}` route because it's defined first, even though `/users/me` is a more specific route.

### Operating System

Linux

### Operating System Details

Operating System Details Ubuntu 20.04

### FastAPI Version

0.108.0

### Pydantic Version

2.5.2

### Python Version

Python 3.10.14

### Additional Context

This issue isn't causing a bug in my current project, but it's something I've run into before and I believe addressing it could improve FastAPI's usability and robustness.

I'm happy to provide any additional details or context if needed. And I'm excited to contribute back to the FastAPI community by helping others with their questions and issues moving forward.

Thanks for considering this suggestion and for all your amazing work on FastAPI!

**Comment by YuriiMotov** [comment]
I also thought that FastAPI could show warning if more specific paths declared after less specific one. It could be an optional feature

**Comment by linkedlist771** [comment]
I think there should be a PR to warn this router configs.
